src/psyclone/psyir/frontend/fparser2.py

```diff
@@ -3274,6 +3277,155 @@ def _if_stmt_handler(self, node, parent):
+        '''
+        Transforms an fparser2 Select_Type_Construct to the PSyIR
+        representation.
+
+        :param node: node in fparser2 tree.
+        :type node: :py:class:`fparser.two.Fortran2003.Select_Type_Construct`
+        :param parent: Parent node of the PSyIR node we are constructing.
+        :type parent: :py:class:`psyclone.psyir.nodes.Node`
+
+        :returns: PSyIR representation of node
+        :rtype: :py:class:`psyclone.psyir.nodes.IfBlock`
+
+        :raises InternalError: If the fparser2 tree has an unexpected \
+            structure.
+        :raises NotImplementedError: If the fparser2 tree contains an \
+            unsupported structure and should be placed in a CodeBlock.
+
+        '''
+        # Check that the fparser2 parsetree has the expected structure
+        if not isinstance(node, Fortran2003.Select_Type_Construct):
+            raise InternalError(
+                f"Failed to find opening select type statement in: {node}")
+        if not isinstance(node.children[-1], Fortran2003.End_Select_Type_Stmt):
+            raise InternalError(
+                f"Failed to find closing select type statement in: {node}")
+
+        # Search for all the TYPE IS and CLASS IS clauses in the
+        # Select_Type_Construct and extract the required
+        # information. This makes for easier code generation later in
+        # the routine.
+        select_idx = -1   # index of the current clause
+        default_idx = -1  # index of the default clause if it exists
+        guard_type = []   # type of guard (class is, type is default class)
+        clause_type = []  # name of the clause
+        stmts = []        # list of statements for each clause
+
+        for idx, child in enumerate(node.children):
+            if isinstance(child, Fortran2003.Select_Type_Stmt):
+                if child.children[0]:
+                    # The selector variable is renamed and this is not
+                    # yet supported.
+                    raise NotImplementedError(
+                        f"The selector variable '{child.children[1]}' is "
+                        f"renamed to '{child.children[0]}' in the select "
+                        f"clause '{str(node)}'. This is not yet supported in "
+                        f"the PSyIR.")
+                selector = child.children[1].string
+            elif isinstance(child, Fortran2003.Type_Guard_Stmt):
+                select_idx += 1
+                guard_type.append(child.children[1])
+                clause_type.append(child.children[0])
+                if child.children[0].lower() == "class default":
+                    default_idx = select_idx
+            elif isinstance(child, Fortran2003.End_Select_Type_Stmt):
+                pass
+            else:
+                # This must be a statement
+                try:
+                    stmts[select_idx].append(child)
+                except IndexError:
+                    stmts.append([])
```

Comment:
I'm not sure what you are expecting here. The else clause must happen after all other clauses have been tested, otherwise `End_Select_Type` will become one of the statements. It may be that the restructuring that I have done means the line numbers you reference are invalid and that it was clearer what you were asking before.

Reply:
The whole implementation has been reworked so I may have addressed this during the changes.